### PR TITLE
	Modifications to try to ensure intellisense offers relevant

### PR DESCRIPTION
**Pull request by rt320**
suggestions when there are multiple tags on a single line and
	when one or more end tags have been removed

	-Logic was introduced to determine if multiple tags are on a
	single line
	-The nearestOpen funtion was modified to take in to account
	multiple tags on a single line
	-The logic for closeElement and cloesElementSlash was modified to
	supply the end tag for previous items whether on a single line
	or multiple lines where the end tag is missing or has been
	deleted. The triggers '/' or '>' should now provide end tags if
	they are missing

	Closes #309

### CODE REVIEW

File: src/language/providers/closeElement.ts [hunk 1]
```diff
@@ -27,98 +32,207 @@ export function getCloseElementProvider() {
         position: vscode.Position
       ) {
         var backpos = position.with(position.line, position.character - 1)
+        var backpos3 = position.with(position.line, position.character - 3)
```
Comment (scholarsmate) [inline src/language/providers/closeElement.ts:35]: We should be using either `let` for mutable variables or `const` for immutable variables.

File: src/language/providers/closeElement.ts [hunk 2]
```diff
@@ -27,98 +32,207 @@ export function getCloseElementProvider() {
...
+                  insertSnippet(
+                    '>\n\t$0\n</' + nsPrefix + 'simpleType>',
+                    backpos
+                  )
+                } else if (
+                  triggerText.endsWith('>') &&
+                  triggerText.includes('dfdl:defineVariable') &&
+                  nearestOpenItem.includes('defineVariable')
+                ) {
+                  var startPos = document.lineAt(position).text.indexOf('<', 0)
```
Comment (scholarsmate) [inline src/language/providers/closeElement.ts:115]: Many / All of the `var` in here should be `const`.

File: src/language/providers/utils.ts [hunk 3]
```diff
@@ -160,7 +274,7 @@ export function checkElementOpen(
       triggerText.includes('<' + nsPrefix + 'element') &&
       !triggerText.includes('</' + nsPrefix + 'element') &&
       !triggerText.includes('/>') &&
-      !triggerText.includes('>')
+      !triggerText.includes('/element>')
```
Comment (scholarsmate) [inline src/language/providers/utils.ts:277]: Is this right?

File: src/language/providers/utils.ts [hunk 4]
```diff
@@ -312,10 +528,13 @@ export function checkBraceOpen(
...
     //.text.substring(0, document.lineAt(lineNum).range.end.character)
 
+    if (!triggerText.includes('{')) {
+      return false
+    }
     if (
       triggerText.includes('"{') &&
       triggerText.includes('}"') &&
-      triggerText.includes('..') &&
+      (triggerText.includes('..') || triggerText.includes('.')) &&
```
Comment (scholarsmate) [inline src/language/providers/utils.ts:537]: This can be simplified to just `triggerText.includes('.')`.

**Comment by stevedlawrence** [comment]
What issue is this resolving? The ticket mentions #309, but that's a pull request and not an issue. Do you mean #165?

As to this change, I have trouble following the logic with all the nested if statemens, which makes me think this might not be best approach. We've been okay with it so far, but it's honestly too much to follow and ensure that we're doing the right thing, and as things get even more complex I don't see this approach being maintainable.

I think maybe a new algorithm is really needed to figure the correct context for autocompletion instead of just a bunch of if-statements. If you want to have a phone call or email thread to discuss some alternative approaches I'm happy to do so.

**Comment by rt320** [comment]
Code fixes will be present in a new PR

**Pull request closed without merging**